### ISSUE TITLE
Auto complete not working for pytorch(1.6.0)

### ISSUE THREAD

**Issue by jwh97nn** [microsoft/pylance-release#418]
## Environment data

-   Language Server version: 2020.9.6
-   OS and version: Ubuntu 18.04
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.6, Anaconda

## Expected behaviour

Auto complete should give all available suggestions(the screenshot is taken under Microsoft Language Server)
![Screenshot from 2020-09-25 17-13-09](https://user-images.githubusercontent.com/16673992/94249962-2ec12600-ff53-11ea-87c1-d647447ac8bf.png)

## Actual behaviour
After switching to Pylance Language Server, auto completion does not seem to work(even manually set python.analysis.extrapaths)
![Screenshot from 2020-09-25 17-12-13](https://user-images.githubusercontent.com/16673992/94249977-3385da00-ff53-11ea-95ab-ea774bde1b03.png)

## Logs

```
[Info  - 5:15:31 PM] Pylance language server 2020.9.6 (pyright c52b78aa) starting
[Info  - 5:15:31 PM] Server root directory: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist
[Info  - 5:15:31 PM] No configuration file found.
[Info  - 5:15:31 PM] Setting pythonPath for service "Desktop": "/home/junwen/anaconda3/envs/pytorch-1.6/bin/python"
[Error - 5:15:31 PM] stubPath /home/junwen/Desktop/typings is not a valid directory.
[Info  - 5:15:31 PM] Assuming Python version 3.6
[Info  - 5:15:31 PM] Assuming Python platform Linux
[Info  - 5:15:32 PM] Searching for source files
[Info  - 5:15:32 PM] Found 1 source file
[Info  - 5:15:32 PM] Background analysis(1) root directory: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist
[Info  - 5:15:32 PM] Background analysis(1) started
Background analysis message: setConfigOptions
Background analysis message: setTrackedFiles
Background analysis message: markAllFilesDirty
Background analysis message: setFileOpened
Background analysis message: analyze
[BG(1)] analyzing: /home/junwen/Desktop/a.py ...
[BG(1)]   parsing: /home/junwen/Desktop/a.py (20ms)
[BG(1)]   parsing: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/2and3/builtins.pyi (170ms)
[BG(1)]   binding: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/2and3/builtins.pyi (55ms)
[BG(1)]   binding: /home/junwen/Desktop/a.py (1ms)
[BG(1)]   checking: /home/junwen/Desktop/a.py ...
[BG(1)]     parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/__init__.py (85ms)
[BG(1)]     binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/__init__.py ...
[BG(1)]       parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_C/__init__.pyi (57ms)
[BG(1)]       binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_C/__init__.pyi (27ms)
[BG(1)]       parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/functional.py (35ms)
[BG(1)]       binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/functional.py (6ms)
[BG(1)]     binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/__init__.py (131ms)
[BG(1)]     parsing: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/2and3/_typeshed/__init__.pyi (4ms)
[BG(1)]     binding: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/2and3/_typeshed/__init__.pyi (2ms)
[BG(1)]     parsing: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/third_party/2and3/typing_extensions.pyi (2ms)
[BG(1)]     binding: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/third_party/2and3/typing_extensions.pyi (1ms)
[BG(1)]     parsing: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/3/typing.pyi (18ms)
[BG(1)]     binding: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/3/typing.pyi (8ms)
[BG(1)]   checking: /home/junwen/Desktop/a.py (269ms)
[BG(1)] analyzing: /home/junwen/Desktop/a.py (517ms)
Background analysis message: resumeAnalysis
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: setFileOpened
Background analysis message: markFilesDirty
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
[FG] parsing: /home/junwen/Desktop/a.py (13ms)
[FG] parsing: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/2and3/builtins.pyi (91ms)
[FG] binding: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/2and3/builtins.pyi (28ms)
[FG] binding: /home/junwen/Desktop/a.py (0ms)
Background analysis message: analyze
[BG(1)] analyzing: /home/junwen/Desktop/a.py ...
[BG(1)]   parsing: /home/junwen/Desktop/a.py (1ms)
[BG(1)]   binding: /home/junwen/Desktop/a.py (0ms)
[BG(1)]   checking: /home/junwen/Desktop/a.py (3ms)
[BG(1)] analyzing: /home/junwen/Desktop/a.py (4ms)
Background analysis message: setFileOpened
Background analysis message: markFilesDirty
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
[FG] parsing: /home/junwen/Desktop/a.py (2ms)
[FG] binding: /home/junwen/Desktop/a.py (1ms)
Background analysis message: setFileOpened
Background analysis message: markFilesDirty
[FG] parsing: /home/junwen/Desktop/a.py (0ms)
[FG] binding: /home/junwen/Desktop/a.py (0ms)
[FG] parsing: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/2and3/_typeshed/__init__.pyi (8ms)
[FG] binding: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/2and3/_typeshed/__init__.pyi (1ms)
[FG] parsing: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/third_party/2and3/typing_extensions.pyi (3ms)
[FG] binding: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/third_party/2and3/typing_extensions.pyi (1ms)
[FG] parsing: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/3/typing.pyi (25ms)
[FG] binding: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/3/typing.pyi (9ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/__init__.py (61ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/__init__.py ...
[FG]   parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_C/__init__.pyi (35ms)
[FG]   binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_C/__init__.pyi (23ms)
[FG]   parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/functional.py (22ms)
[FG]   binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/functional.py (5ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/__init__.py (89ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_utils.py (9ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_utils.py (3ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_utils_internal.py (2ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_utils_internal.py (1ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/version.py (0ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/version.py (0ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_six.py (5ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_six.py (1ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/tensor.py (17ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/tensor.py (3ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/types.py (1ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/types.py (0ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/random.py (3ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/random.py (3ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/serialization.py (28ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/serialization.py (6ms)
[FG] parsing: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/2and3/pickle.pyi (3ms)
[FG] binding: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/2and3/pickle.pyi (1ms)
[FG] parsing: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/2and3/struct.pyi (1ms)
[FG] binding: /home/junwen/.vscode/extensions/ms-python.vscode-pylance-2020.9.6/dist/typeshed-fallback/stdlib/2and3/struct.pyi (0ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_tensor_str.py (7ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_tensor_str.py (6ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/storage.py (3ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/storage.py (1ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_jit_internal.py (11ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_jit_internal.py (3ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_lowrank.py (3ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_lowrank.py (1ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/autograd/__init__.py (2ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/autograd/__init__.py (1ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/autograd/grad_mode.pyi (0ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/autograd/grad_mode.pyi (1ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_ops.py (2ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_ops.py (1ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_classes.py (1ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_classes.py (0ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/multiprocessing/_atfork.py (1ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/multiprocessing/_atfork.py (1ms)
[FG] parsing: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_lobpcg.py (9ms)
[FG] binding: /home/junwen/anaconda3/envs/pytorch-1.6/lib/python3.6/site-packages/torch/_lobpcg.py (3ms)
Background analysis message: analyze
[BG(1)] analyzing: /home/junwen/Desktop/a.py ...
[BG(1)]   parsing: /home/junwen/Desktop/a.py (1ms)
[BG(1)]   binding: /home/junwen/Desktop/a.py (0ms)
[BG(1)]   checking: /home/junwen/Desktop/a.py (1ms)
[BG(1)] analyzing: /home/junwen/Desktop/a.py (2ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: resumeAnalysis
Background analysis message: getDiagnosticsForRange
Background analysis message: setFileOpened
Background analysis message: markFilesDirty
[FG] parsing: /home/junwen/Desktop/a.py (0ms)
[FG] binding: /home/junwen/Desktop/a.py (0ms)
Background analysis message: analyze
[BG(1)] analyzing: /home/junwen/Desktop/a.py ...
[BG(1)]   parsing: /home/junwen/Desktop/a.py (1ms)
[BG(1)]   binding: /home/junwen/Desktop/a.py (0ms)
[BG(1)]   checking: /home/junwen/Desktop/a.py (2ms)
[BG(1)] analyzing: /home/junwen/Desktop/a.py (3ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: resumeAnalysis
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: setFileOpened
Background analysis message: markFilesDirty
[FG] parsing: /home/junwen/Desktop/a.py (0ms)
[FG] binding: /home/junwen/Desktop/a.py (1ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: analyze
[BG(1)] analyzing: /home/junwen/Desktop/a.py ...
[BG(1)]   parsing: /home/junwen/Desktop/a.py (0ms)
[BG(1)]   binding: /home/junwen/Desktop/a.py (0ms)
[BG(1)]   checking: /home/junwen/Desktop/a.py (2ms)
[BG(1)] analyzing: /home/junwen/Desktop/a.py (2ms)
Background analysis message: resumeAnalysis

```

**Comment by jakebailey** [comment]
Odd, pytorch is a very well typed library and you're at the latest version. The logs indicate that its files are being found...

Can you install pytorch into a regular virtual environment and select it and see if it works? I'm half wondering if conda's negatively impacting this, thought I doubt so.

If you were to write something like this:

```python
from torch import random

random.
```

Do you see anything inside `random`?

**Comment by erictraut** [comment]
There was a recent version of pytorch that introduced `if False` conditions for inclusions of its inlined types. They have since fixed this, but I don't know if they have published an updated version. This could be related to that issue.

**Comment by jakebailey** [comment]
If it's recent enough, then they likely haven't. The latest is still 1.6.0 from July (and that's what @IslandCloud is using).

At their 1.6.0 tag, they have stuff like this: https://github.com/pytorch/pytorch/blob/b31f58de6fa8bbda5353b3c77d9be4914399724d/torch/__init__.py#L194

Which I assume is what you're referring to. The latest code uses the correct `TYPE_CHECKING` bool.

**Comment by IslandCloud** [comment]
Thanks for your interest!

So, first I tried the ```random```.
The expected behaviour should be like this(on my Mac without Anaconda installed, and everything worked perfectly):
<img width="1072" alt="Screenshot 2020-09-27 at 15 22 59" src="https://user-images.githubusercontent.com/16673992/94359384-838ea900-00d9-11eb-8e5c-e95f089a5038.png">

The actual behaviour looks like this(on Ubuntu 18.04 with Anaconda installed):
![Screenshot from 2020-09-27 15-59-27](https://user-images.githubusercontent.com/16673992/94359550-92c22680-00da-11eb-87fb-e77c8d27c3df.png)
It seems to work fine, and the outputs exactly matched to that above.

But I found another wired thing, which related to Microsoft Language Server.
When I do the same import using Microsoft Language Server, I got the following output that missing a lot suggestions.
![Screenshot from 2020-09-27 16-02-10](https://user-images.githubusercontent.com/16673992/94359660-42979400-00db-11eb-8fae-d602e3ef75cd.png)

So I did another experiment
The expected behaviour:
<img width="1069" alt="Screenshot 2020-09-27 at 15 48 37" src="https://user-images.githubusercontent.com/16673992/94359867-1a5c6500-00dc-11eb-86be-bb09553f4604.png">

The actual behaviour on Microsoft Language Server(that does not give suggestions to torch.tensor):
![Screenshot from 2020-09-27 15-46-37](https://user-images.githubusercontent.com/16673992/94359968-a1a9d880-00dc-11eb-84da-7c822cd44d38.png)

The actual behaviour on Pylance Language Server(even worse):
![Screenshot from 2020-09-27 15-47-40](https://user-images.githubusercontent.com/16673992/94359911-6ad3c280-00dc-11eb-933a-ea0ddaf538c2.png)

**Comment by IslandCloud** [comment]
I assume it kind of related to this issue [python-language-server #1799](https://github.com/microsoft/python-language-server/issues/1799)

When using Pylance(the second way outputs nothing):
![Screenshot from 2020-09-28 10-51-08](https://user-images.githubusercontent.com/16673992/94385960-fe5ad100-0178-11eb-9c59-2f6070e306e7.png)
![Screenshot from 2020-09-28 10-51-58](https://user-images.githubusercontent.com/16673992/94385967-0286ee80-0179-11eb-80e3-67a1bb535672.png)

When using Microsoft Language Server(both ways did not work well):
![Screenshot from 2020-09-28 10-52-44](https://user-images.githubusercontent.com/16673992/94385981-0dda1a00-0179-11eb-8d88-6a0c856745a7.png)
![Screenshot from 2020-09-28 10-52-57](https://user-images.githubusercontent.com/16673992/94385987-10d50a80-0179-11eb-983d-c7055ab6c242.png)

**Comment by jakebailey** [comment]
There's not really much we can do here without waiting for pytorch to release its most recent fixes. If they are putting their code into `if False` blocks, we're going to see that as dead code and not analyze it. I'm pretty sure the only reason you can get it to work on some other setup is that the pytorch version is older and predates the problem.

**Comment by IslandCloud** [comment]
Exactly, when I downgrade pytorch version to 1.5.0, Pylance works perfectly.
So I will modify the title to specify the pytorch version 1.6.0.
Thanks mate

**Comment by DavidRuhe** [comment]
I'd like to point out that when using Jedi as language server I do get the correct autocompletions.
![image](https://user-images.githubusercontent.com/9398113/95479394-00980900-098b-11eb-8e23-020eb4602d8c.png)

**Comment by jakebailey** [comment]
pytorch 1.7 is now the latest; can you recheck with the updated version to se if this issue is fixed?

**Comment by jwh97nn** [comment]
I've checked Pylance with PyTorch 1.7.0, and it works as expected.

**Comment by jakebailey** [comment]
Thanks for checking. I'm going to close this as there's not much we can do about the code that was published in 1.6 (the `if False` block is pretty devastating) and upgrading the library fixes it.

1.7 shouldn't have this specific issue, though it does have #484 (waiting for a fix in pytorch) and #545 (fixed in the next release).
